tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,728 @@
+      iLabels = tf.reshape(iLabels, tf.constant(newShape)); // flatten one dimension
+      predictions =
+          tf.reshape(
+              predictions,
+              tf.constant(
+                  new long[] {-1L, predictionsShape.size(predictionsShape.numDimensions() - 1)}));
+    }
+
+    @SuppressWarnings("unchecked")
+    Operand<T> loss = tf.nn.sparseSoftmaxCrossEntropyWithLogits(iLabels, predictions);
+    if (updateShape && predictionsRank >= 3) {
+      Shape newShape = predictionsShape.take(predictionsShape.numDimensions() - 1);
+      loss = tf.reshape(loss, tf.constant(newShape));
+    }
+    return loss;
+  }
+
+  /**
+   * Computes the squared hinge loss between labels and predictions.
+   *
+   * <p><code>loss = reduceMean(square(maximum(1 - labels * predictions, 0)))</code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets, values are expected to be -1 or 1. If binary (0 or 1) labels are *
+   *     provided, they will be converted to -1 or 1.
+   * @param predictions the predictions
+   * @param <T> the data type of the predictions and labels
+   * @return the squared hinge loss
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> squaredHinge(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = cast(tf,  labels, dataType);
+    LossTuple<T> lossTuple = LossesHelper.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = lossTuple.getTarget();
+    tLabels = lossTuple.getLabels();
+    Operand<T> one = cast(tf,  tf.constant(1), dataType);
+    Operand<T> zero = cast(tf,  tf.constant(0), dataType);
+
+    tLabels = maybeConvertLabels(tf, tLabels);
+    return tf.math.mean(
+        tf.math.square(tf.math.maximum(tf.math.sub(one, tf.math.mul(tLabels, predictions)), zero)),
+        tf.constant(-1));
+  }
+
+  // private methods/**
+  //   * Calculates the loss
+  //   *
+  //   * @param labels the truth values or labels
+  //   * @param predictions the predictions
+  //   * @param sampleWeights Optional SampleWeights acts as a coefficient for the loss. If a scalar
+  // is
+  //   *     provided, then the loss is simply scaled by the given value. If SampleWeights is a
+  // tensor
+  //   *     of size [batch_size], then the total loss for each sample of the batch is rescaled by
+  // the
+  //   *     corresponding element in the SampleWeights vector. If the shape of SampleWeights is
+  //   *     [batch_size, d0, .. dN-1] (or can be broadcasted to this shape), then each loss element
+  // of
+  //   *     predictions is scaled by the corresponding value of SampleWeights. (Note on dN-1: all
+  // loss
+  //   *     functions reduce by 1 dimension, usually axis=-1.)
+  //   * @param <T> The data type of the predictions, sampleWeights and loss.
+  //   * @param <U> The data type of the labels.
+  //   * @return the loss
+  //   *
```

Comment:
Can we remove this commented-out documentation?

Reply:
OK